### PR TITLE
windows: kill rust-analyzer-proc-macro-srv before deleting stage0 directory

### LINKED ISSUE #107018: Bootstrap fails to wipe the sysroot directory on Windows if proc-macro-server is being used
I tried this:

```rust
x build
```

I expected to see this happen: Bootstrap rebuilds the compiler.

Instead, this happened:
```
PS C:\Users\Joshua Nelson\src\rust> x b clippy --stage 1
Traceback (most recent call last):
  File "C:\Users\Joshua Nelson\src\rust\x.py", line 28, in <module>
    bootstrap.main()
  File "C:\Users\Joshua Nelson\src\rust\src\bootstrap\bootstrap.py", line 937, in main
    bootstrap(help_triggered)
  File "C:\Users\Joshua Nelson\src\rust\src\bootstrap\bootstrap.py", line 912, in bootstrap
    build.download_toolchain()
  File "C:\Users\Joshua Nelson\src\rust\src\bootstrap\bootstrap.py", line 428, in download_toolchain
    shutil.rmtree(bin_root)
  File "C:\Users\Joshua Nelson\AppData\Local\Programs\Python\Python310\lib\shutil.py", line 749, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "C:\Users\Joshua Nelson\AppData\Local\Programs\Python\Python310\lib\shutil.py", line 614, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Users\Joshua Nelson\AppData\Local\Programs\Python\Python310\lib\shutil.py", line 619, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\Joshua Nelson\AppData\Local\Programs\Python\Python310\lib\shutil.py", line 617, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\Joshua Nelson\\src\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\libexec\\rust-analyzer-proc-macro-srv.exe'
```

### Meta

`git merge-base origin/master HEAD` is 9709a438721d679572f0f7c389cd1fa60922fbef

### PR DESCRIPTION
**Pull request by jyn514**
This fixes the following recurring error on windows:
```
Traceback (most recent call last):
  File "C:\Users\jyn\src\rust\x.py", line 29, in <module>
    bootstrap.main()
  File "C:\Users\jyn\src\rust\src\bootstrap\bootstrap.py", line 963, in main
    bootstrap(args)
  File "C:\Users\jyn\src\rust\src\bootstrap\bootstrap.py", line 927, in bootstrap
    build.download_toolchain()
  File "C:\Users\jyn\src\rust\src\bootstrap\bootstrap.py", line 437, in download_toolchain
    shutil.rmtree(bin_root)
  File "C:\Users\jyn\AppData\Local\Programs\Python\Python311\Lib\shutil.py", line 759, in rmtree
    return _rmtree_unsafe(path, onerror)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jyn\AppData\Local\Programs\Python\Python311\Lib\shutil.py", line 617, in _rmtree_unsafe
    _rmtree_unsafe(fullname, onerror)
  File "C:\Users\jyn\AppData\Local\Programs\Python\Python311\Lib\shutil.py", line 622, in _rmtree_unsafe
    onerror(os.unlink, fullname, sys.exc_info())
  File "C:\Users\jyn\AppData\Local\Programs\Python\Python311\Lib\shutil.py", line 620, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\jyn\\src\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\rust-analyzer-proc-macro-srv.exe'
```

Fixes https://github.com/rust-lang/rust/issues/107018.

r? @ChrisDenton

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -434,6 +440,23 @@ def download_toolchain(self):
...
+                        os.path.basename(self.build_dir),
+                        self.build
+                    )
+                    script = (
+                        # NOTE: can't use `taskkill` or `Get-Process -Name` because they error if
+                        # the server isn't running.
+                        'Get-Process | ' +
+                        'Where-Object {$_.Name -eq "rust-analyzer-proc-macro-srv"} |' +
+                        'Where-Object {{$_.Path -match "{}"}} |'.format(regex) +
+                        'Foreach-Object {Stop-Process -Id $_.Id -Force}'
```
Comment (ChrisDenton) [inline src/bootstrap/bootstrap.py:457]: You should be able to simplify this as `Stop-Process` accepts process objects as [input](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/stop-process?view=powershell-7.3#inputs):
```suggestion
                        'Stop-Process -Force'
```

**Review by ChrisDenton — COMMENTED** [review]
This seems reasonable but can we not ask nicely for the server to shutdown cleanly?

**Comment by jyn514** [comment]
What does "shutdown cleanly" mean here? I don't *think* the proc-macro server itself holds any state, it's the main rust-analyzer process that has state.

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -434,6 +440,23 @@ def download_toolchain(self):
...
+                # running. Kill it.
+                if platform_is_win32():
+                    print("Killing rust-analyzer-proc-macro-srv before deleting stage0 toolchain")
+                    regex =  '{}\\\\(host|{})\\\\stage0\\\\libexec'.format(
+                        os.path.basename(self.build_dir),
+                        self.build
+                    )
+                    script = (
+                        # NOTE: can't use `taskkill` or `Get-Process -Name` because they error if
+                        # the server isn't running.
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/bootstrap/bootstrap.py:453]: I'd rather not - being loud on errors caught several bugs as I was working on this (typos in `build` and things like that).

**Comment by jyn514** [comment]
Oh, you just mean remove `-Force` - sure, I can do that.

File: src/bootstrap/bootstrap.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisDenton) [reply src/bootstrap/bootstrap.py:453]: Fair enough

**Comment by ChrisDenton** [comment]
Ok lgtm

@bors r+

**Pull request merged**